### PR TITLE
bootstrap minor improvements and clean-ups

### LINKED ISSUE #128553: compiling rust from github source fails to find rustup
steps..

installed windows 10 on a 64 bit system
installed visual studio 2022 with c++ workload
installed python 3.12
installed git for windows

git clone --depth=1 https://github.com/rust-lang/rust.git
cd rust
python x.py setup

results in c:\source\rust>python x.py setup
downloading https://static.rust-lang.org/dist/2024-07-26/rust-std-beta-x86_64-pc-windows-msvc.tar.xz
################################################################################################################ 100.0%
downloading https://static.rust-lang.org/dist/2024-07-26/rustc-beta-x86_64-pc-windows-msvc.tar.xz
################################################################################################################ 100.0%
downloading https://static.rust-lang.org/dist/2024-07-26/cargo-beta-x86_64-pc-windows-msvc.tar.xz
################################################################################################################ 100.0%
extracting c:\source\rust\build\cache\2024-07-26\rust-std-beta-x86_64-pc-windows-msvc.tar.xz
extracting c:\source\rust\build\cache\2024-07-26\rustc-beta-x86_64-pc-windows-msvc.tar.xz
extracting c:\source\rust\build\cache\2024-07-26\cargo-beta-x86_64-pc-windows-msvc.tar.xz
Building bootstrap
    Updating crates.io index
  Downloaded windows-targets v0.52.6
  Downloaded cpufeatures v0.2.12
  Downloaded heck v0.4.1
  Downloaded itoa v1.0.10
  Downloaded home v0.5.9
  Downloaded filetime v0.2.23
  Downloaded digest v0.10.7
  Downloaded pkg-config v0.3.28
  Downloaded crossbeam-deque v0.8.5
  Downloaded clap_lex v0.6.0
  Downloaded version_check v0.9.4
  Downloaded winapi-util v0.1.6
  Downloaded xz2 v0.1.7
  Downloaded sha2 v0.10.8
  Downloaded semver v1.0.21
  Downloaded quote v1.0.35
  Downloaded junction v1.1.0
  Downloaded anstyle v1.0.4
  Downloaded crossbeam-utils v0.8.20
  Downloaded typenum v1.17.0
  Downloaded unicode-ident v1.0.12
  Downloaded windows-core v0.52.0
  Downloaded serde_derive v1.0.195
  Downloaded serde v1.0.195
  Downloaded memchr v2.7.1
  Downloaded ignore v0.4.21
  Downloaded aho-corasick v1.1.2
  Downloaded cc v1.0.97
  Downloaded serde_json v1.0.111
  Downloaded clap_builder v4.4.12
  Downloaded ryu v1.0.16
  Downloaded syn v2.0.48
  Downloaded opener v0.5.2
  Downloaded toml v0.5.11
  Downloaded tar v0.4.40
  Downloaded object v0.32.2
  Downloaded proc-macro2 v1.0.76
  Downloaded regex-syntax v0.8.2
  Downloaded crossbeam-epoch v0.9.18
  Downloaded bstr v1.9.0
  Downloaded clap_complete v4.4.6
  Downloaded clap v4.4.13
  Downloaded same-file v1.0.6
  Downloaded log v0.4.20
  Downloaded globset v0.4.14
  Downloaded clap_derive v4.4.7
  Downloaded generic-array v0.14.7
  Downloaded walkdir v2.4.0
  Downloaded termcolor v1.4.0
  Downloaded scopeguard v1.2.0
  Downloaded fd-lock v4.0.2
  Downloaded crypto-common v0.1.6
  Downloaded regex-automata v0.4.3
  Downloaded cmake v0.1.48
  Downloaded cfg-if v1.0.0
  Downloaded block-buffer v0.10.4
  Downloaded lzma-sys v0.1.20
  Downloaded libc v0.2.155
  Downloaded windows_x86_64_msvc v0.52.6
  Downloaded winapi v0.3.9
  Downloaded windows-sys v0.52.0
  Downloaded windows v0.52.0
  Downloaded 62 crates (21.7 MB) in 18.71s (largest was `windows` at 11.8 MB)
   Compiling windows_x86_64_msvc v0.52.6
   Compiling proc-macro2 v1.0.76
   Compiling unicode-ident v1.0.12
   Compiling winapi v0.3.9
   Compiling version_check v0.9.4
   Compiling typenum v1.17.0
   Compiling memchr v2.7.1
   Compiling serde v1.0.195
   Compiling cc v1.0.97
   Compiling crossbeam-utils v0.8.20
   Compiling libc v0.2.155
   Compiling pkg-config v0.3.28
   Compiling cfg-if v1.0.0
   Compiling anstyle v1.0.4
   Compiling clap_lex v0.6.0
   Compiling regex-syntax v0.8.2
   Compiling heck v0.4.1
   Compiling log v0.4.20
   Compiling semver v1.0.21
   Compiling serde_json v1.0.111
   Compiling bootstrap v0.0.0 (C:\source\rust\src\bootstrap)
   Compiling cpufeatures v0.2.12
   Compiling itoa v1.0.10
   Compiling scopeguard v1.2.0
   Compiling ryu v1.0.16
   Compiling clap_builder v4.4.12
   Compiling generic-array v0.14.7
   Compiling aho-corasick v1.1.2
   Compiling bstr v1.9.0
   Compiling object v0.32.2
   Compiling windows-targets v0.52.6
   Compiling windows-sys v0.52.0
   Compiling windows-core v0.52.0
   Compiling cmake v0.1.48
   Compiling quote v1.0.35
   Compiling crossbeam-epoch v0.9.18
   Compiling syn v2.0.48
   Compiling crossbeam-deque v0.8.5
   Compiling windows v0.52.0
   Compiling lzma-sys v0.1.20
   Compiling regex-automata v0.4.3
   Compiling crypto-common v0.1.6
   Compiling block-buffer v0.10.4
   Compiling digest v0.10.7
   Compiling sha2 v0.10.8
   Compiling winapi-util v0.1.6
   Compiling opener v0.5.2
   Compiling same-file v1.0.6
   Compiling termcolor v1.4.0
   Compiling walkdir v2.4.0
   Compiling toml v0.5.11
   Compiling globset v0.4.14
   Compiling filetime v0.2.23
   Compiling fd-lock v4.0.2
   Compiling junction v1.1.0
   Compiling home v0.5.9
   Compiling tar v0.4.40
   Compiling ignore v0.4.21
   Compiling clap_derive v4.4.7
   Compiling serde_derive v1.0.195
   Compiling clap v4.4.13
   Compiling clap_complete v4.4.6
   Compiling build_helper v0.1.0 (C:\source\rust\src\tools\build_helper)
   Compiling xz2 v0.1.7
    Finished `dev` profile [unoptimized] target(s) in 1m 53s
downloading https://static.rust-lang.org/dist/2024-07-30/rustfmt-nightly-x86_64-pc-windows-msvc.tar.xz
################################################################################################################ 100.0%
extracting c:\source\rust\build\cache\2024-07-30\rustfmt-nightly-x86_64-pc-windows-msvc.tar.xz to c:\source\rust\build\x86_64-pc-windows-msvc\rustfmt
downloading https://static.rust-lang.org/dist/2024-07-30/rustc-nightly-x86_64-pc-windows-msvc.tar.xz
################################################################################################################ 100.0%
extracting c:\source\rust\build\cache\2024-07-30\rustc-nightly-x86_64-pc-windows-msvc.tar.xz to c:\source\rust\build\x86_64-pc-windows-msvc\rustfmt
Welcome to the Rust project! What do you want to do with x.py?
a) library: Contribute to the standard library
b) compiler: Contribute to the compiler itself
c) tools: Contribute to tools which depend on the compiler, but do not modify it directly (e.g. rustdoc, clippy, miri)
d) dist: Install Rust from source
e) none: Do not modify `config.toml`
Please choose one (a/b/c/d/e): d

To get started, try one of the following commands:
- `x.py dist`
- `x.py build`
`x.py` will now use the configuration at C:\source\rust/src/bootstrap/defaults/config.dist.toml

Rust's CI will automatically fail if it doesn't pass `tidy`, the internal tool for ensuring code quality.
If you'd like, x.py can install a git hook for you that will automatically run `test tidy` before
pushing your code to ensure your code is up to par. If you decide later that this behavior is
undesirable, simply delete the `pre-push` file from .git/hooks.
Would you like to install the git hook?: [y/N] y
Linked `src/etc/pre-push.sh` to `.git/hooks/pre-push`

Command "rustup" "--version" (failure_mode=Exit) did not execute successfully.
It was not possible to execute the command: Error { kind: NotFound, message: "program not found" }

Build completed unsuccessfully in 0:04:31

### PR DESCRIPTION
**Pull request by onur-ozkan**
3rd commit fixes https://github.com/rust-lang/rust/issues/128553#issuecomment-2266119948.

### CODE REVIEW

File: src/bootstrap/src/core/builder.rs [hunk 1]
```diff
@@ -1449,15 +1449,11 @@ impl<'a> Builder<'a> {
             assert_eq!(target, compiler.host);
         }
 
-        if self.config.rust_optimize.is_release() {
-            // FIXME: cargo bench/install do not accept `--release`
-            // and miri doesn't want it
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/builder.rs:1454]: This doesn't seem like a FIXME note; it's more like a regular note.

File: src/bootstrap/src/core/build_steps/tool.rs [hunk 2]
```diff
@@ -1091,18 +1091,13 @@ macro_rules! tool_extended {
...
 tool_extended!((self, builder),
     Cargofmt, "src/tools/rustfmt", "cargo-fmt", stable=true;
     CargoClippy, "src/tools/clippy", "cargo-clippy", stable=true;
     Clippy, "src/tools/clippy", "clippy-driver", stable=true, add_bins_to_sysroot = ["clippy-driver", "cargo-clippy"];
     Miri, "src/tools/miri", "miri", stable=false, add_bins_to_sysroot = ["miri"];
     CargoMiri, "src/tools/miri/cargo-miri", "cargo-miri", stable=true, add_bins_to_sysroot = ["cargo-miri"];
-    // FIXME: tool_std is not quite right, we shouldn't allow nightly features.
-    // But `builder.cargo` doesn't know how to handle ToolBootstrap in stages other than 0,
-    // and this is close enough for now.
-    Rls, "src/tools/rls", "rls", stable=true, tool_std=true;
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/build_steps/tool.rs:1105]: This is an outdated FIXME we had for rust-demangler.

File: src/bootstrap/src/core/download.rs [hunk 3]
```diff
@@ -616,8 +616,6 @@ impl Config {
         };
 
         // For the beta compiler, put special effort into ensuring the checksums are valid.
-        // FIXME: maybe we should do this for download-rustc as well? but it would be a pain to update
-        // this on each and every nightly ...
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/download.rs:620]: It sounds painful and I think we will never be doing this.

**Comment by onur-ozkan** [comment]
Rebased.

File: src/bootstrap/src/core/download.rs (hunk 3 above)
Comment (albertlarsan68) [reply src/bootstrap/src/core/download.rs:620]: For a future possibility, maybe distribute the hashes with the tarballs?
Or resurrect the signing of the artifacts...

**Review by albertlarsan68 — APPROVED** [review]
LGTM

**Pull request merged**